### PR TITLE
Reworking test infrastructure

### PR DESCRIPTION
**Pull request by ericsagnes**
Updated the PR description and title on 2016/09/03 according to current state.
### Motivation for this change

This PR has the following goals:
- make Nix tests easier to work with (`callTest` and `nix-run-test`)
- Allow module maintainers to have better control of module tests by tying the tests to the modules, and putting the test file in a folder of their liking.
### Main improvements
- Simpler test infrastructure, tests are just a function
- Tests are not tied to `make-test.nix` anymore, and can be run from anywhere
- Relationship between tests and modules, maintainership is improved as tests and modules can be in the same/close folders
- `nix-run-test` command
- moved `callTest` to `lib`, so it can easily be called from custom projects
### Changes
- Adds a new `meta.tests` attribute to modules
- Add a new `callTest` function that has a wider use range than the previous one
- Removes previous `callTest`, `callSubTests` as their functionality is superseded by the new `callTest`
- Move `callTest` to `lib.testing` so it can easily be called from other places than `release.nix`
- Move all module related tests to the module `meta.tests` attributes
- Modules related tests are in `tests.module` subset of `release.nix`
- Simplify test files by removing the need of calling `makeTest` in test files, test files should be a test set, a set of test sets, or a test derivation
- Some test related files that were in `nixos/lib`, namely `test-driver/*` and `testing.nix`, are relocated in `lib/testing`
- A new `nix-run-test` command had been added to run test in an easy way
- Manual has been updated accordingly to the changes
- Removes unneeded `nixos/tests/make-test.nix`
### Possibly controversial changes
- Removes previous `callTest`, `callSubTests` as their functionality is superseded by the new `callTest`
- Some test related files that were in `nixos/lib`, namely `test-driver/*` and `testing.nix`, are relocated in `lib/testing`
- Removes unneeded `nixos/tests/make-test.nix`
#### modules

`meta.tests` is a new attribute of modules that takes a set of tests.
Key is the set name, value is the test file, eg:

```
  meta.tests = {
    containers-ipv4       = ./tests/containers-ipv4.nix;
    containers-ipv6       = ./tests/containers-ipv6.nix;
    containers-bridge     = ./tests/containers-bridge.nix;
    containers-imperative = ./tests/containers-imperative.nix;
    containers-extra_veth = ./tests/containers-extra_veth.nix;
  };
```
#### release.nix

In `release.nix`, module tests are gathered with the `moduleTests` function, and added in the `tests.module` set.
Problematic tests can be blacklisted by adding their name to the `blacklistedModuleTests` list.

``` nix
blacklistedModuleTests = [ "postgis" ];
```

Module tests can be run like any other tests, eg:

```
$ nix-build -A tests.module.panamax ./nixos/release.nix
```

The `moduleTests` accept a set argument to set arguments on dedicated tests if needed.

```
  tests.module = moduleTests {
    cadvisor = { systems = [ "x86_64-linux" ]; };
    kubernetes = { systems = [ "x86_64-linux" ]; };
    docker = { systems = [ "x86_64-linux" ]; };
    dockerRegistry = { systems = [ "x86_64-linux" ]; };
    etcd = { systems = [ "x86_64-linux" ]; };
    fleet = { systems = [ "x86_64-linux" ]; };
    panamax = { systems = [ "x86_64-linux" ]; };
  };
```
#### callTest

`callTest` has been reworked to accept test sets, set of tests sets or test derivations.
Also, there is no need to use `makeTest` in the test files anymore.

A new `lib` subset, `lib.testing` was added. `lib.testing` expose only the `callTest` function.

This change make `callTest` a part of `lib`, so it can be called from places other than `release.nix` , examples:

a. Call a test:

``` nix
lib.testing.callTest ./nixos/tests/simple.nix {};
```

b. Call a test for only x86_64-linux:

``` nix
lib.testing.callTest ./nixos/tests/simple.nix { systems = [ "x86_64-linux" ]; };
```

c. Call a test with custom parameters:

``` nix
lib.testing.callTest ./nixos/tests/nfs.nix { version = 3; };
```
#### nix-run-test

`nix-run-test`, a new utility to run nix tests has been added. It is just a wrapper around `nix-build` and `callTest` to run tests in an easy manner.

a. Call a test:

``` shell
$ nix-run-test ./nixos/tests/simple.nix
```

b. Prepare an interactive test:

``` shell
$ nix-run-test --interactive ./nixos/tests/simple.nix
```
### TODO
- Test in a Hydra
###### Things done
- [ ] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by aszlig** [comment]
Cool, this was something on my TODO list for quite a while.

I already have an implementation of this in vuizvui for machine-dependant tests:

https://github.com/openlab-aux/vuizvui/blob/master/modules/core/tests.nix

Having tests in within the meta attributes however won't help for machine-related tests (I'm building channels based on [which settings are enabled in the machine configuration](https://headcounter.org/hydra/channel/custom/openlab/vuizvui/channels.machines.aszlig.tishtushi)) so maybe it's a better idea to add them as a module option.

Which benefit does it have if the tests are defined in the meta attributes?

**Comment by ericsagnes** [comment]
@aszlig 

Thanks for looking at this!
I added a few commits to clean up and give a better idea of what it should look like when finished.

> Which benefit does it have if the tests are defined in the meta attributes?

The benefits I can see of having the tests in meta are:
- obvious one, tests are tied to modules
- it is easy to fetch all the modules tests in release.nix (see `moduleTests` function in `release.nix`)
- there is no need to declare a new option in every module
- not really because of meta, but as a side effect tests and modules are closer on the filesystem

> Having tests in within the meta attributes however won't help for machine-related tests

To be honest I haven't thought at all about machine-related tests for this. 
It is true meta just tie the tests with a module file and knows nothing about the options defined in the module, so it cannot be a great help for machine-related tests.

One approach I can think of (no idea if it is realistic or not), would be to define a function `relevant :: Config -> Bool` in the test, so by passing `config` to it, we can decide if the test is relevant for a particular config.
Then it would just be a matter of collecting all the "relevant" tests for a particular config.

Test pseudo code:

```nix
{ pkgs, ... } :

{
  name = "foo";

  relevant = config:
    config.services.httpd.enable && elem "proxy_balancer" config.services.httpd.extraModules;

  nodes = {
    # ...
  };

  testScript = ''
    ...
  '';
}
```

`release.nix` pseudo code:

```nix
  dummyAttrs = { ... };
  moduleTests = (import lib/eval-config.nix { modules = []; }).config.meta.tests;
  relevantTests = config: filterAttrs (k: v: (v dummyAttrs).relevant config) moduleTests;
  testsForMachineA = relevantTests machineAConfig;
```

**Comment by ericsagnes** [comment]
Update: Added a few commits and updated the PR description.

**Comment by ericsagnes** [comment]
**Update**

The PR is almost complete functionality wise, from now I am just planning to test and fix any problems that could rise without changing much of its logic. 

It changed a lot since its creation so please take the time to read the updated description.

As this contains some changes that could be considered controversial (see description for a list) and touch a wide range of files, I am very interested in getting some feedback. Especially from @edolstra and @aszlig as they are the main contributors to the current test infrastructure and @domenkozar and @joachifm as they were interested by the concept.

Thank you!

**Comment by aszlig** [comment]
> One approach I can think of (no idea if it is realistic or not), would be to define a function `relevant :: Config -> Bool` in the test, so by passing `config` to it, we can decide if the test is relevant for a particular config.

The downside of this is that once modules are refactored it's very likely that people forget to adapt the test definitions.

In `vuizvui` the tests are defined by a module option so defining a test in a module's config is tying it to the module, [like this](https://github.com/openlab-aux/vuizvui/blob/4fca46cfcfd9c80f8beb76d053e157d5c31ddcd4/modules/programs/gnupg/default.nix#L104).

Of course this has the downside that it's no longer that easy to enumerate *all* tests, except by using a dummy config that enables all modules with tests.

Another thing I'm interested is having NixOS tests for normal packages, like for example for Chromium. Possibly also tied to whether such a package is defined in `systemPackages`. The implementation for this could be done by mapping through `systemPackages` for packages that define a `meta.test` attribute.

> - there is no need to declare a new option in every module

Why would one need to declare an option for a test?

**Comment by ericsagnes** [comment]
> The downside of this is that once modules are refactored it's very likely that people forget to adapt the test definitions.

If that breaks tests for some, they will fix the definitions.
But yes, it is not a perfect solution.

> Another thing I'm interested is having NixOS tests for normal packages, like for example for Chromium. Possibly also tied to whether such a package is defined in systemPackages. The implementation for this could be done by mapping through systemPackages for packages that define a meta.test attribute.

Actually, now that tests files can be moved around easily, I was thinking to move packages related tests in packages meta and closer to packages files in a second phase.

> Why would one need to declare an option for a test?

Sorry, I was a little fast on reading your config. No, there is no need for that.

Anyway, the original goal of this PR was to tie modules and tests together so modules and tests maintainership can be improved.
I think machine-dependent testing would be best addressed in a future PR. 
Maybe we can create an issue to discuss about the right way to implement it?

But let me know if you think that some changes this PR brings would make a future implementation machine-dependent testing difficult, I would be very happy to fix these.

> Also, I'd rename nix-run-tests to nixos-run-tests, because those tests are NixOS-specific.

I was hesitating on this point, `nix-run-test` is just a wrapper to `nix-build` so I hoped it could work for packages tests on platforms were nix only is installed, such as other linuxes, maybe that was too much of optimistic thinking?

**Comment by ericsagnes** [comment]
update:
- rebased on master to fix conflicts
- moved mongodb test to the module

File: nixos/modules/services/networking/avahi-daemon.nix [hunk 1]
```diff
@@ -204,4 +204,8 @@ in
 
   };
 
+  meta.tests = {
+    "avahi" = ./tests/avahi.nix;
```
Comment (kampfschlaefer) [inline nixos/modules/services/networking/avahi-daemon.nix]: This is inconsistent with the rest of the attributes where the test-name is without quotation.

File: nixos/modules/services/networking/cjdns.nix [hunk 2]
```diff
@@ -263,4 +263,8 @@ in
 
   };
 
+  meta.tests = {
+    "cjdns" = ./tests/cjdns.nix;
```
Comment (kampfschlaefer) [inline nixos/modules/services/networking/cjdns.nix]: This is also inconsistent with the rest of the attributes where the test-name is without quotation.

File: nixos/modules/services/x11/desktop-managers/gnome3.nix [hunk 3]
```diff
@@ -190,5 +190,9 @@ in {
 
   };
 
+  meta.tests = {
+    "gnome3" = ./tests/gnome3.nix;
```
Comment (kampfschlaefer) [inline nixos/modules/services/x11/desktop-managers/gnome3.nix]: Here the quotation is also not needed...

**Review by kampfschlaefer — CHANGES_REQUESTED** [review]
I like this in general.

Found three small inconsistencies, maybe fix them?

**Comment by edolstra** [comment]
This mostly looks good, except that I don't think we should move the NixOS testing stuff to the non-NixOS part of Nixpkgs. They're NixOS-specific after all.

We should probably also stop adding functions that are not packages to `pkgs`.

I don't think we need `nix-run-test`. IMHO it doesn't add anything over `nix-build`.

**Comment by ericsagnes** [comment]
@kampfschlaefer Thanks for the review and for noticing these! I will fix that.

@edolstra 

> This mostly looks good, except that I don't think we should move the NixOS testing stuff to the non-NixOS part of Nixpkgs. They're NixOS-specific after all.

I agree with this. Just to explain the motivation behind moving test functions to lib, it was to be able to easily call tests in custom projects having release.nix like files (right now this is not trivial as the test functions are in defined in `release.nix`).

What about making changes so `<nixpkgs/nixos>` exposes its `lib`.  As a result tests could be called with something like `with import <nixpkgs/nixos> {}; lib.callTest ./some-test.nix`?
Would you approve such changes?

> I don't think we need nix-run-test. IMHO it doesn't add anything over nix-build.

No problems, I will remove it and add documentation on how to run tests with `nix-build` instead.
The motivation was just to save some typing for the lazy, but I agree this is slightly an overkill.

**Comment by ericsagnes** [comment]
update:
- rebase on master to fix conflicts
- include samba test

@edolstra 

Asking again as you might have missed my previous comment.

What about making changes so `<nixpkgs/nixos>` exposes its `lib`. As a result tests could be called with something like with `import <nixpkgs/nixos> {}; lib.callTest ./some-test.nix`?
Would you approve such changes?

**Comment by grahamc** [comment]
FWIW: I really like this change. I will miss being able to look in nixos/tests to see all of them, but I do like it.

**Comment by edolstra** [comment]
Hm, I missed the fact that this turns tests into modules. Not sure if that's an improvement, because currently you can just do `nix-build nixos/tests/simple.nix -A test`, which won't work on modules.

**Comment by ericsagnes** [comment]
> Hm, I missed the fact that this turns tests into modules. Not sure if that's an improvement, because currently you can just do nix-build nixos/tests/simple.nix -A test, which won't work on modules.

It does not really turn tests into modules, it just ties the tests to their related modules with the `meta.tests` attribute.

There are a few ways to run the tests tied to modules with this PR:
- The usual `nix-build` + `release.nix` attribute path: 
    
    ```
    nix-build -A tests.module.panamax ./nixos/release.nix
    ```

- using `nix-run-test` + path to the test file: 

    ```
    nix-run-test nixos/tests/simple.nix
    ```
- using `nix-build` + path to the test file

    ```
    nix-build -E "with import <nixpkgs/lib/testing> {}; callTest ./nixos/tests/simple.nix"
    ```

The third notation is slightly long to type, so that motivated the introduction of `nix-run-test`.
On the other side, this PR introduce the ability to run a test file located in any directory. (current tests have a relative import to `make-test.nix` that make them very hard to move)

**Comment by 0xABAB** [comment]
@ericsagnes Does it just smell funny or is it dead?

**Comment by Ekleog** [comment]
@ericsagnes Are you still willing to push this forward? If not, given how badly this appears to have bitrotten, maybe it would make sense to open an issue with the changes you tried to do (from a quick read, replace in-file `make-test.nix` with out-of-file `callTest` and addition of `meta.tests` to modules), and close this PR?

**Comment by mmahut** [comment]
Are there any updates on this pull request, please?

**Comment by davidak** [comment]
Instead of `meta.tests` we have `passthru.tests` now.

https://github.com/NixOS/nixpkgs/pull/44439
https://github.com/NixOS/nix/issues/2532#issuecomment-487411058

The testing situation has also changed lastly. Module tests have been rewritten in Python. Package tests are possible now.

https://github.com/NixOS/nixpkgs/issues/73076

@ericsagnes can you create individual issues or PRs for the changes that are still missing?

**Pull request closed without merging**